### PR TITLE
Fix table navigation commands in Google Docs

### LINKED ISSUE #9494: error in google docs table nav
### Steps to reproduce:
1. open a new docs document with braille mode enabled.
2. create a 3x3 table.
3. fill out the table.
3. try to press `alt+ctrl+<arrow keys>`

### Actual behavior:

```
Input: kb(laptop):alt+control+downArrow
ERROR - scriptHandler.executeScript (12:43:20.176):
error executing script: <bound method Dynamic_EditableTextWithAutoSelectDetectionEditorIa2WebIAccessible.script_nextRow of <NVDAObjects.Dynamic_EditableTextWithAutoSelectDetectionEditorIa2WebIAccessible object at 0x0380B150>> with gesture u'alt+ctrl+down arrow'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 187, in executeScript
  File "documentBase.pyc", line 174, in script_nextRow
  File "documentBase.pyc", line 161, in _tableMovementScriptHelper
  File "documentBase.pyc", line 126, in _getNearestTableCell
TypeError: coercing to Unicode: need string or buffer, int found
```

### Expected behavior:

NVDA can move around the table.

#### NVDA installed/portable/running from source:
instaled

#### NVDA version:
alpha-17077,7f274eca

#### Windows version:
10
#### Name and version of other software in use when reproducing the issue:
docs, chrome, firefox
#### Have you tried any other versions of NVDA? If so, please report their behaviors.

worked before 2019.1

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #9494

### Summary of the issue:
Currently trying to move around a table in Google Docs with Firefox or Chrome, using NVDA's table navigation commands fails with  an exception.
Specifically, NVDAObject's rowNumber, columnNumber etc expose web author specified rowindex and colindex ARIA values which don't work with the other IAccessible2 table methods for looking up cells.
For browseMode at least, this was already partly addressed by exposing special table-physical* attributes on controlFields which were used for table navigation, where as presentation (speech, braille) used the standard table-* attributes.

### Description of how this pull request fixes the issue:
Rewrite how presentational verses physical table properties are exposed entirely. Instead exposing default / physical table properties as their standard rowNumber, columnNumber etc. If the underlying API does have presentational coordinates, then expose these as the presentational* values, using them only for presentation, and not physical table navigation.
Specific changes:
* Expose the real IAccessible2 table coordinates for navigation via rowNumber, columnNumber, rowCount and columnCount NvDAObject properties, and controlField attributes, rather than IA2-specific table properties.
* Add new optional properties to NVDAObjects: presentationalRowNumber, presentationalColumnNumber, presentationalRowCount and presentationalColumnCount.
* Make use of these presentational* properties in speech if they are implemented, otherwise fall back to the standard properties.
* Provide implementations of the presentational* properties for tables on NVDAObjects for IAccessible2 that uses rowindex and colindex IA2 aria attributes, allowing web authors to override how table coordinates are presented without confusing actual table navigation logic.
* Similarly for controlFields, expose table-* and table-*-presentational attributes allowing for overriding of table coordinate presentation when the web author has specified it.
* IA2TextInfo's updateSelection method: If the requested start and end offsets are the same (I.e. the caller wants a collapsed selection) call setCaretOffset instead. This gets around strange bugs in Google Chrome / Google Docs where making collapsed selections in table cells selects the entire cell.

### Testing performed:
Using this testcase: 
[table.html.txt](https://github.com/nvaccess/nvda/files/3171353/table.html.txt)
Opened it in both Mozilla Firefox and Google Chrome.
Ensured that:
* Arrowing into the first table, and then switching to focus mode, that NVDA's table navigation commands (control/alt+up/down/left/right) successfully moved around the table.
*  Ensured that when arrowing through the second table in browseMode, that table coordinates reported correctly. (The table is a standard table with no ARIA properties). Also ensured that table navigation commands still worked.
* When arrowing into the 3rd table, ensured that table coordinates reported as expected. Although this table is only 3 rows by 2 columns, it should report as 20 rows by 20 columns, columns should be 10 and 11, and rows should be numbered: 1, 10 and 11. These coordinates have been specified using aria-rowindex etc.
* Created a table in Google Docs with Braille mode on, and ensured that  NVDA's table navigation commands successfully move around the table.

### Known issues with pull request:
None.

### Change log entry:
Bug fixes:
* It is again possible to use NvDA's table navigation commands in Google Docs, in Firefox and Chrome.

### CODE REVIEW

File: source/NVDAObjects/__init__.py [hunk 1]
```diff
@@ -558,12 +558,36 @@ def _get_rowNumber(self):
 		"""
 		raise NotImplementedError
 
+	def _get_presentationalRowNumber(self):
+		"""
+		An optional version of the rowNumber property used purely for speech and braille presentation if implemented.
```
Comment (leonardder) [inline source/NVDAObjects/__init__.py]: Some of the lines in here seem to have a pretty long length. I recall @feerrenrut preferring a length of at most 120, preferably around 80 characters.

File: source/NVDAObjects/__init__.py [hunk 2]
```diff
@@ -578,12 +602,36 @@ def _get_rowCount(self):
...
+		This is never used for navigational logic.
+		This property should be implemented if the table has virtual content which may not all be loaded at one time.
+		For example, a table with 1000 rows and 1000 columns, yet the table only shows perhaps 10 rows by 10 columns at a time.
+		Although the  columnCount might be 10, the user needs to  be told the table really has 1000 columns. 
+		If the underlying APIs do not distinguish between virtual and physical cell coordinates, then this property should not be implemented.
+		@rtype: int
+		"""
+		raise NotImplementedError
+
 	def _get_rowSpan(self):
```
Comment (leonardder) [inline source/NVDAObjects/__init__.py]: Just to make sure, could there also be something like a presentational row/column span?

File: source/speech.py [hunk 3]
```diff
@@ -1237,7 +1250,9 @@ def getControlFieldSpeech(attrs,ancestorAttrs,fieldType,formatConfig=None,extraD
 		containerContainsText=_("with %s items")%childControlCount
 	elif fieldType=="start_addedToControlFieldStack" and role==controlTypes.ROLE_TABLE and tableID:
 		# Table.
-		return " ".join((nameText,roleText,stateText, getSpeechTextForProperties(_tableID=tableID, rowCount=attrs.get("table-rowcount"), columnCount=attrs.get("table-columncount")),levelText))
+		rowCount=(attrs.get("table-rowcount-presentational") or attrs.get("table-rowcount"))
+		columnCount=(attrs.get("table-columncount-presentational") or attrs.get("table-columncount"))
+		return " ".join((nameText,roleText,stateText, getSpeechTextForProperties(_tableID=tableID, rowCount=rowCount, columnCount=columnCount),levelText))
```
Comment (leonardder) [inline source/speech.py]:
```suggestion
		return " ".join((
			nameText,
			roleText,
			stateText,
			getSpeechTextForProperties(
				_tableID=tableID,
				rowCount=rowCount,
				columnCount=columnCount
			),
			levelText
		))
```

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 4]
```diff
@@ -815,21 +803,21 @@ VBufStorage_fieldNode_t* GeckoVBufBackend_t::fillVBuf(IAccessible2* pacc,
...
 
 	// Add some presentational table attributes
-	// Note these are only for reporting, the physical table attributes (table-physicalrownumber etc) for aiding in navigation etc are added  later on.
+	// Note these are only for reporting, the physical table attributes (table-rownumber etc) for aiding in navigation etc are added  later on.
 	// propagate table-rownumber down to the cell as Gecko only includes it on the row itself
 	if(parentPresentationalRowNumber) 
-		parentNode->addAttribute(L"table-rownumber",parentPresentationalRowNumber);
+		parentNode->addAttribute(L"table-rownumber-presentational",parentPresentationalRowNumber);
 	const wchar_t* presentationalRowNumber=NULL;
 	if((IA2AttribsMapIt = IA2AttribsMap.find(L"rowindex")) != IA2AttribsMap.end()) {
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: While here, I would love it if you could change these conditions into two statements:
```suggestion
	IA2AttribsMapIt = IA2AttribsMap.find(L"rowindex");
	if(IA2AttribsMapIt != IA2AttribsMap.end()) {
```

File: nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp [hunk 5]
```diff
@@ -113,18 +113,16 @@ static IAccessible2* IAccessible2FromIdentifier(int docHandle, int ID) {
 template<typename TableType> inline void fillTableCounts(VBufStorage_controlFieldNode_t* node, IAccessible2* pacc, TableType* paccTable) {
 	wostringstream s;
 	long count = 0;
-	// Fetch row and column counts and add them as two sets of attributes on this vbuf node.
+	// Fetch row and column counts and add them as attributes on this vbuf node.
 	// The first set: table-physicalrowcount and table-physicalcolumncount represent the physical topology of the table and can be used programmatically to understand table limits.
```
Comment (feerrenrut) [inline nvdaHelper/vbufBackends/gecko_ia2/gecko_ia2.cpp]: Yes, this comment is no longer relevant.

File: source/NVDAObjects/IAccessible/__init__.py [hunk 6]
```diff
@@ -1144,18 +1154,23 @@ def _get_cellCoordsText(self):
 			rowText=self.rowNumber
 		return "%s %s"%(colText,rowText)
 
-	def _get_columnNumber(self):
+	def _get_presentationalColumnNumber(self):
 		index=self.IA2Attributes.get('colindex')
+		try:
```
Comment (feerrenrut) [inline source/NVDAObjects/IAccessible/__init__.py]: `_get_presentationalRowNumber` also tries the parent, where this doesn't. Why is it different?
The rest of this code looks the same, I suggest extracting it for re-use.

File: source/documentBase.py [hunk 7]
```diff
@@ -169,6 +165,7 @@ def _tableMovementScriptHelper(self, movement="next", axis=None):
 		speech.speakTextInfo(info,formatConfig=formatConfig,reason=controlTypes.REASON_CARET)
 		info.collapse()
 		self.selection = info
+		#info.updateCaret()
```
Comment (feerrenrut) [inline source/documentBase.py]: remove this or add some comments about why we specifically don't want to do this here.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/NVDAObjects/IAccessible/__init__.py (hunk 6 above)
Comment (michaelDCurran) [reply source/NVDAObjects/IAccessible/__init__.py]: We expect that row and column numbers are exposed on cells. But some implementations may only expose rowIndex on the row object (parent of the cells) itself. Tables never expose column objects though thus we never need to worry about that. 
I'm not sure abstracting the rest of that code makes it any more readable. I'm struggling to work out what you would actually call it.

**Comment by michaelDCurran** [comment]
@feerrenrut I have addressed the rest of the review comments from you and @leonardder

**Review by feerrenrut — APPROVED** [review]

**Comment by feerrenrut** [comment]
I don't think it should block this PR, however there is still a question from @leonardder:

> Just to make sure, could there also be something like a presentational row/column span?

File: source/NVDAObjects/__init__.py (hunk 2 above)
Comment (michaelDCurran) [reply source/NVDAObjects/__init__.py]: I'm not aware of any colspan or rowspan ARIA attributes at this stage. Only rowIndex and rowCount. Spanning would also have an affect on the physical table thus I don't think that could ever be virtualized.

**Pull request merged**